[FLINK-3231][streaming-connectors] FlinkKinesisConsumer rework to handle Kinesis resharding

tzulitai opened the pull request:
This change attempts to solve 2 issues:
1. [FLINK-3231](https://issues.apache.org/jira/browse/FLINK-3231): Handle Kinesis-side resharding.
2. [FLINK-4020](https://issues.apache.org/jira/browse/FLINK-4020): Remove shard list querying from Kinesis consumer constructor.

Some notes on the implementation:
- All subtasks has a thread that continuously polls for changes in the Kinesis stream, and uses exponential backoff with jitter to try to even out the concurrent Kinesis client describeStream operations across subtasks. Continuously polling is necessary because there's currently no way to "signal" a subtask that it has a new shard it should be subscribing to.
- A big change is that all subtasks run a fetcher that continues to poll for shards, even if the subtask initially didn't have shards to consume (before, a MAX_VALUE waterwark was sent out).
- The `describeStream` operation doesn't fail after 3 retries now. Instead, if the failures it was getting from Kinesis was only `LimitExceededException`, we continue to retry (with jitter backoff) until we succeed.
- Apart from the unit tests, I've manually tested this with `ManualExactlyOnceWithStreamReshardingTest`. However, since the `FlinkKinesisProducer` currently has some problem that records are resent when Kinesis streams are resharded (thus not allowing the exactly-once test to pass at all), this manual test uses a normal event generator instead of a producer topology running the `FlinkKinesisProducer`.

Since this PR introduces considerable rework on the Kinesis consumer, I'll wait until this can be merged before submitting [FLINK-4080](https://issues.apache.org/jira/browse/FLINK-4080) & [FLINK-4019](https://issues.apache.org/jira/browse/FLINK-4019).

tzulitai commented:
The ManualExactlyOnceTest and ManualExactlyOnceWithStreamReshardingTest is a bit buggy, fixing with a follow-up commit ...

tzulitai commented:
@rmetzger tagging just to make sure you're notified of this PR :)
When will you be free to review? Just for my own time allocation for when to continue working on the Kinesis connector. If there's anything majorly wrong with the implementation explained in the above comment, please let me know and I'll try to address them before effort on a detailed review.
Thanks in advance!

rmetzger commented:
I'll review your change now.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,156 +17,481 @@
+		while (running) {
+			// once either shutdownFetcher() or stopWithError()
+			// is called, we will escape this loop
+		}
```
This loop is extremely inefficient. Is a very busy wait loop (one CPU core will constantly burn cycles on this loop).

tzulitai replied:
Yup this is really bad ... will fix!

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/IKinesisProxy.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.Map;
+
+public interface IKinesisProxy {
```
The `I` prefix is not commonly used within Flink.
I would suggest `KinesisProxyInterface` as a name here.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -86,32 +141,28 @@ public KinesisProxy(Properties configProps) {
 	 * @return the batch of retrieved records
 	 */
-	public GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) {
+	public GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) throws InterruptedException {
```
Missing `@Override` annotation

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -122,29 +173,36 @@ public GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) {
 	 */
-	public List<KinesisStreamShard> getShardList(List<String> streamNames) {
-		List<KinesisStreamShard> shardList = new ArrayList<>();
+	public GetShardListResult getShardList(List<String> streamNames) throws InterruptedException {
```
Missing `@Override` annotation

rmetzger replied:
We can probably drop this unused method.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -122,29 +173,36 @@ public GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) {
+	 * @param streamNamesWithLastSeenShardIds a map with stream as key, and last seen shard id as value
+	 * @return shard list result
+	 */
+	public GetShardListResult getShardList(Map<String,String> streamNamesWithLastSeenShardIds) throws InterruptedException {
```
Missing `@Override` annotation

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/IKinesisProxy.java:
```diff
@@ -0,0 +1,31 @@
+
+public interface IKinesisProxy {
+	GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) throws InterruptedException;
+	GetShardListResult getShardList(List<String> streamNames) throws InterruptedException;
```
This method is unused.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -159,42 +217,65 @@ public String getShardIterator(KinesisStreamShard shard, String shardIteratorTyp
 	 * Get metainfo for a Kinesis stream, which contains information about which shards this Kinesis stream possess.
 	 *
+	 * This method is using a "full jitter" approach described in
+	 * <a href="http://google.com">https://www.awsarchitectureblog.com/2015/03/backoff.html</a>. This is necessary
```
Why is this linking to google.com?

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -86,32 +141,28 @@ public KinesisProxy(Properties configProps) {
-				} catch (InterruptedException interruptEx) {
-					//
+				if (seed == null) {
+					seed = new Random();
```
Is there a reason for creating a new RNG with every `getRecords` call?

tzulitai replied:
Nope. I'll change to create it as a static class member. `getRecords` and `describeStream` can actually share the random when calling `fullJitterBackoff`.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -206,4 +287,9 @@ private DescribeStreamResult describeStream(String streamName, String startShard
 			throw new RuntimeException("Stream is not Active or Updating");
```
Can you put the actual stream status into the exception?

tzulitai replied:
Actually, should we really be throwing a RuntimeException if a single stream is not in ACTIVE or UPDATING status? Say the consumer is to fetch 3 streams, and only 1 is found to be CREATING / DELETING. Perhaps we should treat this case the same as "can only find shards for some of the streams", and a warning log here will do?

There's a check that stops the consumer in the new `ShardDiscoverer` if no shards can be initially found at all to read from anyway.

rmetzger replied:
I agree! We shouldn't fail hard here.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -159,42 +217,65 @@ public String getShardIterator(KinesisStreamShard shard, String shardIteratorTyp
-			remainingRetryTimes--;
 		}
 
 		if (streamStatus == null) {
```
The `RuntimeException` below has a typo.
`Can't get stream info from ____ after`.

Also, I wonder where the number 3 is coming from.

tzulitai replied:
Actually, since we'll be retrying until we get a describeStreamResult now, this message and RuntimeException can be removed.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/KinesisStreamShard.java:
```diff
@@ -61,19 +68,23 @@ public String getShardId() {
 	public String getStartingSequenceNumber() {
-		return shard.getSequenceNumberRange().getStartingSequenceNumber();
+		SequenceNumberRange sequenceNumberRange = shard.getSequenceNumberRange();
+		return (sequenceNumberRange == null) ? null : sequenceNumberRange.getStartingSequenceNumber();
```
In what cases is the sequence range null?

tzulitai replied:
I added these null checks because in the tests, fake `KinesisStreamShards` may be created without giving them a SequenceNumberRange / HashNumberRange etc. if it wasn't relevant to the test. Perhaps I should revert these null check changes, and always give all values for the fake `KinesisStreamShard`s?

tzulitai replied:
The `Shard`s returned from actual Kinesis will never have null values for these.

rmetzger replied:
If this is the case, than I don't see the reason for these checks.
In general, I have the feeling that the `KinesisStreamShard` is overly complicated. Maybe its easier to just have a method getShard() instead of proxying everything?

tzulitai replied:
Okay, I agree!

rmetzger commented:
I'm not done with the review yet. I hope I find some time over the weekend to continue with it.

tzulitai commented:
@rmetzger Thank you! No problem, I can wait and address the current comments meanwhile.

rmetzger commented:
Okay, sounds good.

tzulitai commented:
Hi @rmetzger ,
I've addressed your initial comments with the last commit. No urgent hurry on the remaining review, please take your time :) Thanks!

tzulitai commented:
I've also tested the last commit on the unit & manual tests.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyInterface.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Map;
+
+public interface KinesisProxyInterface {
```
Can you add a little bit of javadocs to the class and the methods?

tzulitai replied:
Sure, no problem.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardDiscoverer.java:
```diff
@@ -0,0 +1,215 @@
+	 * @param initialState the initial state to assign to each new shard that this subtask should subscribe to
+	 */
+	private void discoverNewShardsAndSetInitialStateTo(String initialState) throws InterruptedException {
+		GetShardListResult shardListResult = kinesis.getShardList(new HashMap<>(streamToLastSeenShard));
```
Why are you creating a new HashMap here?

tzulitai replied:
Mistake, absolutely unnecessary. Thanks for noticing!

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,156 +17,489 @@
+			shardDiscovererAndSubscriberExecutor.submit(new ShardSubscriber<>(this));
+
+			// after startup, we wait util either shutdownFetcher() or stopWithError() is called
+			fetcherShutdownLock.wait();
```
I think the `shutdownFetcher()` can never call notifyAll because this method is never releasing the lock on `fetcherShutdownLock`.

What happens instead is that Flink is interrupting the thread holding the lock. This interrupt will cause an InterruptedException, which is not catched at all.

I don't think this waiting mechanism is necessary here at all. Maybe we can use the `awaitTermination()` for that (but maybe there is an even better solution available ... I'll continue with the review)

tzulitai replied:
Doesn't calling `wait()` release the lock on `fetcherShutdownLock`?

rmetzger replied:
You are right. Sorry.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,156 +17,489 @@
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} is starting the shard subscriber ...", runtimeContext.getIndexOfThisSubtask());
+			}
+			shardDiscovererAndSubscriberExecutor.submit(new ShardSubscriber<>(this));
```
Why do we need a separate thread for this?

tzulitai replied:
Perhaps we can simply integrate the functionality of this thread into `ShardDiscoverer`? Once new shards we should be subscribing to is discovered, we open up a new `ShardConsumer` for it. This we also won't need the `pendingShards` queue.

rmetzger replied:
I wonder if we can use the main thread for the shard discovery? Its sitting there waiting only.
This way, we could get rid of the `shardDiscovererAndSubscriberExecutor` thread pool.

tzulitai replied:
Good idea! Yes, I think we can use the main thread. I'll probably integrate the work done in `ShardDiscoverer` and `ShardSubscriber` into `KinesisDataFetcher#runFetcher()`. `KinesisDataFetcher#shutdownFetcher()` should also be much simpler with this change.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardDiscoverer.java:
```diff
@@ -0,0 +1,215 @@
+ * {@link KinesisDataFetcher#pendingShards} queue with initial state, i.e. where in the new shard we should start
+ * consuming from.
+ */
+public class ShardDiscoverer<T> implements Runnable {
```
It seems that this thread is continuously querying the Kinesis APIs. The only way to slow these requests down is by the jittered backoff.
I think we should query the kinesis api for reshards at a configurable frequency, probably by default only every 30 seconds or so.

tzulitai replied:
You are right, this thread continuously polls Kinesis. I was worried about the delay it would introduce, and how users would take the delay with it being a streaming source.
But yes I agree it will be best to introduce configurable frequency, I was initially a bit unsure of the continuous polling myself. We can add description about this delay of discovering new shards after resharding in the consumer documentation.

rmetzger replied:
Okay. I'm still not done with the review, so my statement might not make much sense:
We could also trigger a special shard discovery when a reading thread hits a deleted / unavailable shard. This way, we can reduce the delay for our users.

tzulitai replied:
Some feedback for this approach:
At first I had thought of the solution you mentioned. The problem was that the new shards will not necessarily belong to the subtask that hit a closed shard (according to the new way we are assigning shards to subtasks in `ShardDiscoverer#isShouldSubscribeTo()`), so we still need other subtasks to continuously poll Kinesis in order for the right subtasks to pick up the new shards. Simply put, I don't think we can soley rely on the event of encountering a closed shard to let the new shards get fully picked up by the correct subtasks.

We could drop the assigning of `ShardDiscoverer#isShouldSubscribeTo()`, and simply let subtasks be responsible for new child shards of a closed shard it was previously consuming. However, with this approach we'll be subject to severe load imbalance, because in Kinesis, users can freely choose which shards to split and merge. To counter this we will a need rebalance mechanism, as well as a "merged snapshot" on restore on failure.

rmetzger replied:
I agree with
> Simply put, I don't think we can soley rely on the event of encountering a closed shard to let the new shards get fully picked up by the correct subtasks.

My idea was to trigger an additional discovery when we hit a closed shard.
We could for example discover shards every minute OR when we hit a closed shard. But you are right, this won't solve the issue with other parallel instances.
Lets keep it as is for now.
Thank you for the explanation.

Lets keep the assignment from `ShardDiscoverer#isShouldSubscribeTo()`.

tzulitai replied:
@rmetzger double check: are we still adding configurable frequency with 30 second default for this?

rmetzger replied:
Yes.

rmetzger commented:
@tzulitai: I'm still not done with the review. I hope I can get it done in the next two hours.

tzulitai commented:
@rmetzger 
No problem, thank you! I'll reply to your comments after you finish reviewing :)

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -86,32 +144,25 @@ public KinesisProxy(Properties configProps) {
-			i++;
 		}
 
 		if (getRecordsResult == null) {
```
Not sure if throwing an exception here is a good idea.
From the javadocs of `getRecord()` it says: "Note that if the shard has been closed, the shard iterator can't return more data and <a>GetRecords</a> returns <code>null</code>". This means that our code would fail each time some shards are closed?

tzulitai replied:
Misunderstanding here.
The `getRecordsResult` will only be `null` here if we fail to get any response from Kinesis due to `ProvisionedThroughputExceededException` even after 3 retries (default setting of `KinesisConfigConstants.DEFAULT_SHARD_GETRECORDS_RETRIES`). We'll always get a `GetRecordsResult` regardless of the status of the shard.

What the javadoc is saying is that if we reached the end of a shard (the shard is closed) `GetRecordsResult#getNextShardIterator()` will return a `null`, meaning that we won't be able to continue consuming the shard. In `ShardConsumer`, we always renew the `nextShardIterator` with the value returned in `GetRecordsResult`. If we get a `null`, we know the shard is closed and terminate the `ShardConsumer` thread. So, no more `getRecords()` calls will be made to the shard after it is closed.

tzulitai replied:
My reasoning for failing hard here and not in the `describeStream()` operation for limit / provision throughput exceeded exceptions is that I think rate exceeding for `getRecords()` on a single shard may be cause by numerous reasons that the user should handle.
1. The user might have other non-Flink concurrent consumers reading the same shard (Kinesis limits to 5 concurrent reads per shard.) The user should either slow down the non-Flink consumers, or slow down our `ShardConsumer` by adjusting the `KinesisConfigConstants.CONFIG_SHARD_GETRECORDS_IDLE_MILLIS` we will be introducing in https://github.com/apache/flink/pull/2071.
2. The size of each record is too large, and the user needs to adjust `KinesisConfigConstants.CONFIG_SHARD_GETRECORDS_MAX` to a smaller value.

`describeStream()` on the other hand is an internal operation we need to continuously do, and make sure we get results for the connector to work.

rmetzger replied:
Thank you for the good explanation. I agree to keep it as it is.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:
```diff
@@ -155,55 +199,79 @@ public GetRecordsResult getRecords(String shardIterator, int maxRecordsToGet) {
+	@Override
 	public String getShardIterator(KinesisStreamShard shard, String shardIteratorType, String startingSeqNum) {
-		return kinesisClient.getShardIterator(shard.getStreamName(), shard.getShardId(), shardIteratorType, startingSeqNum).getShardIterator();
+		return kinesisClient.getShardIterator(shard.getStreamName(), shard.getShard().getShardId(), shardIteratorType, startingSeqNum).getShardIterator();
```
@rmetzger This method doesn't do the `ProvisionedThroughputExceededException` check / retries like the `getRecords()` method. Perhaps we should be checking that + fail hard after retrying too, as it also has a 5 requests per second limit? Reasoning is the same as my reply to your comment on the `getRecords()` method.

rmetzger replied:
Yes, I think we should have retries there as well. In particular during start up, many parallel Flink instances will query AWS for the shard iterator.

tzulitai replied:
Okay. Actually, the 5 rps limit for `getShardIterator()` is per shard. We'll only have one thread making this call for each shard, so the rate limit won't be a problem for us. But as explained above, we should still retry + fail hard to let the user know if they need to adjust other non-Flink consumers.

rmetzger commented:
I'm done with my first full review. I'll test the code tomorrow.
Let me know if you need help addressing my comments.

tzulitai commented:
@rmetzger
Thanks a lot for your helpful review :) I should be able to address your comments before tomorrow so we can start testing it.

tzulitai commented:
I'll rebase this after #2180 is merged to reflect in this PR the exactly-once fix, and commit altogether with the changes to address the remaining comments.

rmetzger commented:
Okay, I'll merge the exactly once fix now.

tzulitai commented:
@rmetzger Thanks Robert!
Sorry for the delay, got caught up with other things. Will push within the next 2~3 hrs.

tzulitai commented:
@rmetzger
Just pushed the changes to address comments + rebase on the exactly-once fix / user-agent fix.
Sorry, had some trouble with the rebasing and took more time than I expected.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,157 +17,553 @@
+				try {
+					Thread.sleep(discoveryIntervalMillis);
+				} catch (InterruptedException iex) {
+					// the sleep may be interrupted by shutdownFetcher()
```
I'm a bit concerned about this part here, it seems a bit floppy. I'm not sure if there's a better way to avoid start sleeping for the discovery interval when `running` is set to false by `shutdownFetcher()` mid-loop ...
@rmetzger Please let me know what you think about this one.

rmetzger replied:
I think the approach is fine. running will be set to false, then the interrupt to leave the sleep.
Flink will also start interrupting the thread if its not reacting.
Also, once the next sleep has finished it will check the running flag again, leaving the loop.

tzulitai replied:
> Flink will also start interrupting the thread if its not reacting.
I see! Okay, thanks, let's leave it as it is.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/KinesisConfigConstants.java:
```diff
@@ -68,10 +97,30 @@
+	public static final double DEFAULT_SHARD_GETITERATOR_BACKOFF_EXPONENTIAL_CONSTANT = 1.5;
 
-	public static final int DEFAULT_SHARD_RECORDS_PER_GET = 100;
+	public static final long DEFAULT_SHARD_DISCOVERY_INTERVAL_MILLIS = 10000L;
```
I'm using 10s for default discovery interval here. I tested it and the originally suggested 30s seemed a bit too long as a default, IMHO. Can change it back to 30s if you think it's more appropriate.

rmetzger replied:
Okay. 10 seconds is okay as well.

rmetzger commented on docs/apis/streaming/connectors/kinesis.md:
```diff
@@ -60,10 +60,10 @@ to setup Kinesis streams. Make sure to create the appropriate IAM policy and use
 
 ### Kinesis Consumer
```
I think we should also mention the threading model of the Kinesis consumer in the documentation.

Users should know that each parallel Flink instance will constantly query Kinesis for shards. So if a user has 5 shards, but runs Flink with a parallelism of 50, there will be 50 threads constantly querying Kinesis.

Also, we should explain that there will always be one thread per shard running.

tzulitai replied:
Good point! I'll add this.

rmetzger commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,157 +17,553 @@
+			KinesisStreamShardState seededShardState = subscribedShardsState.get(seededStateIndex);
+
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} will start consuming seeded shard {} from sequence number {} with ShardConsumer {}" +
```
I think the + at the end of the string is wrong.

tzulitai replied:
Indeed :)

rmetzger commented:
I tested the code, also with a shard-merging:
`aws kinesis merge-shards  --shard-to-merge shardId-000000000001 --adjacent-shard-to-merge shardId-000000000002 --stream-name flink-test`

and everything worked nicely, the log statements were good.
If you want to improve it a little bit, we should maybe log at debug level each time we discover new shards (just to show that everything is working as expected).

The only thing missing are some minor documents, then, I think we are good to merge.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -17,157 +17,553 @@
+					LOG.info("Subtask {} has discovered a new shard {} due to resharding, and will start consuming " +
+							"the shard from sequence number {} with ShardConsumer {}",
+						runtimeContext.getIndexOfThisSubtask(), newShardState.getKinesisStreamShard().toString(),
+						newShardState.getLastProcessedSequenceNum(), newStateIndex);
```
@rmetzger We're already logging here when new shards are discovered. Should I still add the debug level log you mentioned, perhaps in `KinesisDataFetcher#discoverNewShardsToSubscribe()`?

rmetzger replied:
I would add a LOG.debug statement everytime we call `discoverNewShardsToSubscribe()`.

tzulitai replied:
Oh I see. Ok!

tzulitai commented:
@rmetzger Thank you for your review. I hope I've addressed your last comments with the last commit.

For the documentation, I added a bit more apart from the threading model: 1) enabling checkpointing, mostly borrowed from the Kafka documentation, and 2) information on how the consumer internally uses the Kinesis APIs so that users can make sense of any limit warnings they are getting in the logs.

Please let me know if there's anything else to address!

rmetzger commented:
Thank you for the quick fix. I hope I can take a look tonight. Otherwise, I'll look at it early next week.
Thanks a lot for addressing my comments to quickly.

rmetzger commented:
I'm going to merge this once travis gives me green light ;)
